### PR TITLE
Log function results

### PR DESCRIPTION
**Pull request by isaacaka**
High Order Function for logging input and corresponding output of matlab functions.

Usage guide:

Some metadata about the functions and data needs to be passed to the higher order function so files and directories can be created with the appropriate names. `Varagin` does not store variable names passed in so this must also be passed as metadata.

**Add Metadata**
```
metadata = struct();
metadata.siteFile = 'name_of_site_file
metadata.oneFluxDir = '/path/to/ONEFlux/directory';
metadata.relArtifactsDir = '/relative/path/to/tests/test_artifacts';
metadata.frequency = 10  %frequncy_to_log_input/ouput, defualt is 10 if not specified
```
The site file name can be extracted as so:
```
cSiteYrSplit = strsplit(cSiteYr, '.');
metadata.siteFile = cSiteYrSplit{1};

```
cpdFindChangePoint does not take cSitreYr in as an argument, so could be passed in using varargin.

**This**    

```
metadata.inputNames = {'t_it_', 'NEE_it_', 'updated_uStar_it_', 'T_it_', 'fNight_it_', 'fplot', 'cSiteYr'};
metadata.outputNames = {'xCp2', 'xStats2', 'xCp3', 'xStats3'};

[xCp2,xStats2, xCp3,xStats3] = ...
	logFuncResult('log.json', @cpdEvaluateUStarTh4Season20100901, metadata, 					 
                 t(it),NEE(it),updated_uStar(it),T(it),fNight(it),fPlot,cSiteYr);
```

**replaces this**
```
[xCp2,xStats2, xCp3,xStats3] = ...
	cpdEvaluateUStarTh4Season20100901 ...
		(t(it),NEE(it),updated_uStar(it),T(it),fNight(it),fPlot,cSiteYr); 
```

In the main **launch.m** function at the very end you must clear the global frequency cache:

`clear global globalFrequencyMetadata
`

--------------------------------
**Using the function will give something in the format:**

tests/{function_name}_artifacts/site_name/
-- Data1
-- Data2
etc

**Which looks like this:**
 
Screenshot from 2024-11-27 17-11-44

**The log file looks like this:**

Screenshot from 2024-11-27 17-51-29

### CODE REVIEW

**Review by dorchard — APPROVED** [review]
LGTM! Thanks. Do we have code that is using this already as well?

**Comment by dorchard** [comment]
@isaacaka to put an example of this / a brief bit of documentation in our hackmd notes.

**Comment by dorchard** [comment]
We want to store multiple iterations' outputs, so we need a little global state (a struct, mapping function names to iteration count) to record the number of times something has been logged so far (`logFuncResult` will need increment each call that occurs), which we can then use to take a sample (e.g., every 5th iteration) into the files. @j-emberton did something similar before (if you could dig out that would be great).

**Comment by isaacaka** [comment]
Updated to add the functionality

**Comment by dorchard** [comment]
Looks good- will approve once I've used it succesfully myself.

File: oneflux_steps/ustar_cp_refactor_wip/logFuncResult.m [hunk 1]
```diff
@@ -0,0 +1,180 @@
+function varargout = logFuncResult(filename, f, metadata, varargin)
+    % LOGFUNCRESULT Logs function inputs and outputs, saving each variable as a CSV or JSON file.
+    %
+    %   [outputs...] = LOGFUNCRESULT(filename, f, varargin)
```
Comment (dorchard) [inline oneflux_steps/ustar_cp_refactor_wip/logFuncResult.m]: Doesn't match the head of the function which also includes `metadata`

File: oneflux_steps/ustar_cp_refactor_wip/logFuncResult.m [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    %
+    %   Inputs:
+    %       - filename: Name of the log file (e.g., 'log.json').
+    %       - f: Function handle to the function you want to log.
+    %       - varargin: Cell array of input arguments for the function f.
+    %
```
Comment (dorchard) [inline oneflux_steps/ustar_cp_refactor_wip/logFuncResult.m:11]: Missing the `metadata` input in the documentation here.

**Comment by dorchard** [comment]
Is there any advice on where to put the `metadata`? Can it be somewhat global? What is the best way to feed in the `siteFile` field for this given that this will be dynamic?

**Comment by isaacaka** [comment]
I normally put it at the top of each of the main functions which have been refactored. That would be _cpdAssign, cpdBootstrap, cpdEvaluate, cpdFindChangePoint and launch_
I extract the site data from `cSiteYr` which contains the file name.
e.g.
```
cSiteYrSplit = strsplit(cSiteYr, '.');
metadata.siteFile = cSiteYrSplit{1};

```
cpdFindChangePoint does not take cSitreYr in as an argument, so could be passed in using varargin.

**Comment by dorchard** [comment]
Working for me now, just a couple of documentation comments then we can merge this. Probably putting an example 
 usage and`metadata` into the documentation string here would also be good rather than having this just in the PR.

File: oneflux_steps/ustar_cp_refactor_wip/logFuncResult.m [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    %   Outputs:
+    %       - varargout: The outputs returned by the function f. 
+
+
```
Comment (dorchard) [inline oneflux_steps/ustar_cp_refactor_wip/logFuncResult.m:48]: Add example metadata and usage here too (from the PR)

File: oneflux_steps/ustar_cp_refactor_wip/logFuncResult.m (hunk 1 above)
Comment (dorchard) [reply oneflux_steps/ustar_cp_refactor_wip/logFuncResult.m]: I've added this now

File: oneflux_steps/ustar_cp_refactor_wip/logFuncResult.m (hunk 2 above)
Comment (dorchard) [reply oneflux_steps/ustar_cp_refactor_wip/logFuncResult.m:11]: Added

**Comment by dorchard** [comment]
I've added a `metadata.offset` so that multiple call sites can be logged at different phases rather than overlapping. I've also moved documentation from this PR into the comment. I think we can merge this now.

**Pull request merged**